volumes/vds/py/vds_functions.py

```diff
@@ -0,0 +1,464 @@
+        LOGGER.critical(f"Error fetching from {table_name}.")
+        LOGGER.critical(exc)
+        con.close()
+
+def fetch_and_insert_data(select_conn, insert_conn, select_query, insert_query, table_name, batch_size = 100000):
+    #generic function to pull and insert data using different connections and queries.
+    batch=1
+    try:
+        with select_conn.get_conn() as con:
+            with con.cursor() as cur:
+                LOGGER.info(f"Fetching {table_name}")
+                cur.execute(select_query)
+                data = cur.fetchmany(batch_size)
+                while not len(data) == 0:
+                    LOGGER.info(f"Batch {batch} -- {len(data)} rows fetched from {table_name}.")
+                    insert_data(insert_conn, insert_query, table_name, data)
+                    data = cur.fetchmany(batch_size)
+                    batch = batch + 1
+    except Error as exc:
+        LOGGER.critical(f"Error fetching from {table_name}.")
+        LOGGER.critical(exc)
+        con.close()
+
+def insert_data(conn, query, table_name, data):
+    #generic function to insert data
+    try:
+        with conn.get_conn() as con:
+            with con.cursor() as cur:                
+                # Insert cleaned data into the database
+                LOGGER.info(f"Inserting into {table_name}.")
+                execute_values(cur, query, data, page_size = 1000) #31s for 800k rows of vdsvehicledata w/10000. --41s w/1000. >3 minutes w/100
+                LOGGER.info(f"Inserted {len(data)} rows into {table_name}.")
+    except Error as exc:
+        LOGGER.critical(f"Error inserting into {table_name}.")
+        LOGGER.critical(exc)
+        con.close()
+
+def pull_raw_vdsdata(rds_conn, itsc_conn, start_date):
+#pulls data from ITSC table `vdsdata` and inserts into RDS table `vds.raw_vdsdata`
+
+    # Pull raw data from Postgres database
+    raw_sql = sql.SQL("""SELECT 
+        divisionid,
+        vdsid,
+        timestamputc, --timestamp in INTEGER (UTC)
+        lanedata
+    FROM public.vdsdata
+    WHERE
+        timestamputc >= extract(epoch from timestamp with time zone {start}) :: INTEGER
+        AND timestamputc < extract(epoch from timestamp with time zone {start} + INTERVAL '1 DAY') :: INTEGER
+        AND divisionid = 2 --other is 8001 which are traffic signal detectors and are mostly empty.
+        AND length(lanedata) > 0; --these records don't have any data to unpack.
+    """).format(
+        start = sql.Literal(start_date + " 00:00:00 EST5EDT")
+    )
+
+    insert_query = sql.SQL("""INSERT INTO vds.raw_vdsdata (
+                                        division_id, vds_id, datetime_20sec, datetime_15min, lane, speed_kmh, volume_veh_per_hr, occupancy_percent
+                                        ) VALUES %s;""")
+    batch_size = 100000
+    
+    #pull data in batches and transform + insert.
+    try:
+        with itsc_conn.get_conn() as con:
+            with con.cursor() as cur:
```

Comment:
you can combine these `with`s to reduce the nesting levels like this:
`with itsc_conn.get_conn() as con, con.cursor() as cur:`